torch/optim/optimizer.py

```diff
@@ -720,44 +768,27 @@ def load_state_dict(self, state_dict: StateDict) -> None:
             if hook_result is not None:
                 state_dict = hook_result
 
-        # Validate the state_dict
-        groups = self.param_groups
-
         # Deepcopy as we write into saved_groups later to update state
         saved_groups = deepcopy(state_dict['param_groups'])
 
+        # Validate the state_dict
+        groups = self.param_groups
         if len(groups) != len(saved_groups):
             raise ValueError("loaded state dict has a different number of "
                              "parameter groups")
-        param_lens = (len(g['params']) for g in groups)
-        saved_lens = (len(g['params']) for g in saved_groups)
-        if any(p_len != s_len for p_len, s_len in zip(param_lens, saved_lens)):
-            raise ValueError("loaded state dict contains a parameter group "
-                             "that doesn't match the size of optimizer's group")
```

Comment:
I removed the checks for this chunk above because we check it during the cast hook, but it may be safest to do the checks in both places.

Another option is to not check this in the hook.

Reply:
I guess there are two possible points here:
- Having it only here allows you to make sure this test will always run even if the user removes the cast hook and it will validate that what the user hooks generated is a valid state dict.
- Having it in the hook allows the user to "remove it" if they want this "broken" behavior for some reason.
- Have both so that we guarantee that both pre-hooks and post-hooks always generate valid state dict

Which one do you think is preferred here?

Reply:
1. You're assuming we give users the handle to remove the preregistered hook 😛 which I don't yet. Thoughts on storing the returned handle as an attribute on the Optimizer class?
2. I am between either giving users freedom vs. having invariants on our state_dicts. Right now I am leaning towards having the checks in both to maintain the invariants. Do you have more thoughts?

Reply:
Given the current state of state dict, I would agree that for now, more checks are better than less. So I would vote on having it in both as well.